client/scala/src/test/scala/ai/verta/client/entities/TestExperimentRun.scala

```diff
@@ -31,80 +32,172 @@ class TestExperimentRun extends FunSuite {
+   *  @param getter function to retrieve a single piece of metadata
+   *  @param metadataName type of the metadata
+   */
+  def testAlreadyLogged(
+    logger: ExperimentRun => (String, ValueType) => Try[Unit],
+    multiLogger: ExperimentRun => Map[String, ValueType] => Try[Unit],
+    getter: ExperimentRun => String => Try[Option[ValueType]],
+    metadataName: String
+  ) = {
     val f = fixture
 
     try {
-      f.expRun.logMetric("existing", 0.5)
-      val logAttempt = f.expRun.logMetric("existing", 0.5)
+      logger(f.expRun)("existing", 0.5)
+      val logAttempt = logger(f.expRun)("existing", 0.5)
       assert(logAttempt.isFailure)
-      assert(logAttempt match {case Failure(e) => e.getMessage contains "Metric being logged already exists"})
+      assert(logAttempt match {case Failure(e) => e.getMessage contains f"${metadataName} being logged already exists"})
 
-      val logAttempt2 = f.expRun.logMetrics(Map("existing" -> 0.5, "other-metric" -> 0.3))
+      val logAttempt2 = multiLogger(f.expRun)(Map("existing" -> 0.5, "other" -> 0.3))
       assert(logAttempt2.isFailure)
-      assert(logAttempt2 match {case Failure(e) => e.getMessage contains "Metric being logged already exists"})
-      assert(f.expRun.getMetric("other-metric").get.isEmpty)
+      assert(logAttempt2 match {case Failure(e) => e.getMessage contains f"${metadataName} being logged already exists"})
+      assert(getter(f.expRun)("other").get.isEmpty)
     } finally {
       cleanup(f)
     }
   }
 
-  test("getMetric should return None when a non-existing key is passed") {
+  /** Test function to verify that the map interface of a metadata works
+   *  @param getMap function to get the map interface of metadata
+   *  @param getter function to retrieve a single piece of metadata
+   */
+  def testMapInterface(
+    getMap: ExperimentRun => () => mutable.Map[String, ValueType],
+    getter: ExperimentRun => String => Try[Option[ValueType]]
+  ) = {
     val f = fixture
 
     try {
-      assert(f.expRun.getMetric("some-metric").get.isEmpty)
+      val map = getMap(f.expRun)()
+      map += ("some" -> 0.5)
+      assert(map.get("some").get.asDouble.get == 0.5)
+      assert(getter(f.expRun)("some").get.get.asDouble.get equals 0.5)
+      assert(map.get("non-existing").isEmpty)
     } finally {
       cleanup(f)
     }
   }
 
+  test("getMetric should retrieve the correct logged metric") {
+    testMetadata(_.logMetric, _.logMetrics, _.getMetric, _.getMetrics)
+  }
+
+  test("logMetric(s) should fail when pass an existing key") {
+    testAlreadyLogged(_.logMetric, _.logMetrics, _.getMetric,  "Metric")
+  }
+
+  test("getMetric should return None when a non-existing key is passed") {
+    testNonExisting(_.getMetric)
+  }
+
   test("metrics map should behave like other metric methods") {
+    testMapInterface(_.metrics, _.getMetric)
+  }
+
+  test("getAttribute(s) should retrieve the correct logged attributes") {
+    // interface is a bit different
```

Comment:
What is different? The support for strings?

Reply:
Yeah the support for passing a list of attributes key to the get all metadata method. It's literally only one assert call different :(

Reply:
Acutally I might be able to refactor this one too using optional parameter.

Reply:
See if it looks better.